File: sdk/java/src/main/java/com/xxxxteam/tempmail/providers/Anonbox.java
package com.xxxxteam.tempmail.providers;

import com.xxxxteam.tempmail.Email;
import com.xxxxteam.tempmail.EmailInfo;
import com.xxxxteam.tempmail.HttpClient;
import com.xxxxteam.tempmail.HttpResult;
import com.xxxxteam.tempmail.Normalizer;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Anonbox 渠道 — https://anonbox.net (CCC)
 * GET /en/ 解析 HTML 获取邮箱，HTTP mbox 格式取信。
 */
public final class Anonbox {

    private static final String CHANNEL = "anonbox";
    private static final String PAGE_URL = "https://anonbox.net/en/";
    private static final String BASE = "https://anonbox.net";

    private static final Pattern MAIL_LINK = Pattern.compile(
            "<a href=\"https://anonbox\\.net/([a-z0-9-]+)/([A-Za-z0-9._~-]+)\">https://anonbox\\.net/[^\"]+</a>",
            Pattern.CASE_INSENSITIVE);
    private static final Pattern DD_RE = Pattern.compile("(?is)<dd([^>]*)>([\\s\\S]*?)</dd>");
    private static final Pattern DISPLAY_NONE = Pattern.compile("display\\s*:\\s*none", Pattern.CASE_INSENSITIVE);
    private static final Pattern P_RE = Pattern.compile("(?is)<p>([\\s\\S]*?)</p>");
    private static final Pattern SPAN_RE = Pattern.compile("(?is)<span\\b[^>]*>[\\s\\S]*?</span>");
    private static final Pattern TAG_RE = Pattern.compile("<[^>]+>");
    private static final Pattern EXPIRES_RE = Pattern.compile(
            "(?is)Your mail address is valid until:</dt>\\s*<dd><p>([^<]+)</p>");

    private Anonbox() {
    }

    private static Map<String, String> htmlHeaders() {
        Map<String, String> h = new LinkedHashMap<>();
        h.put("User-Agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 "
                + "(KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36");
        h.put("Accept", "text/html,application/xhtml+xml");
        return h;
    }

    private static Map<String, String> mboxHeaders() {
        Map<String, String> h = new LinkedHashMap<>();
        h.put("User-Agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 "
                + "(KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36");
        h.put("Accept", "text/plain,*/*");
        return h;
    }

    private static String stripTags(String html) {
        String s = TAG_RE.matcher(html).replaceAll("");
        return s.replace("&nbsp;", " ").replace("&amp;", "&")
                .replace("&lt;", "<").replace("&gt;", ">").trim();
    }

    /**
     * 创建临时邮箱。
     *
     * @return 邮箱信息
     */
    public static EmailInfo generate() {
        HttpResult r = HttpClient.get(PAGE_URL, htmlHeaders());
        r.ensureSuccess();
        String[] parsed = parseEnPage(r.getBody());
        String email = parsed[0];
        String token = parsed[1];
        String exp = parsed[2];
        return new EmailInfo(CHANNEL, email, token, null, exp);
    }

    /**
     * 获取邮件列表（mbox 格式解析）。
     *
     * @param token 路径令牌 "{inbox}/{secret}"
     * @param email 邮箱地址
     * @return 邮件列表
     */
    public static List<Email> getEmails(String token, String email) {
        if (token == null || token.isBlank()) {
            throw new IllegalArgumentException("anonbox: token 不能为空");
        }
        String path = token.endsWith("/") ? token : token + "/";
        HttpResult r = HttpClient.get(BASE + "/" + path, mboxHeaders());
        if (r.getStatusCode() == 404) return new ArrayList<>();
        r.ensureSuccess();

        String raw = r.getBody().trim();
        if (raw.isEmpty()) return new ArrayList<>();

        // 按 mbox 分割
        String[] blocks = raw.split("(?m)(?=^From )");
        List<Email> emails = new ArrayList<>();
        for (String block : blocks) {
            block = block.trim();
            if (block.isEmpty()) continue;
            Map<String, Object> rawMap = mboxBlockToRaw(block, email);
            emails.add(Normalizer.normalizeEmail(rawMap, email));
        }
        return emails;
    }

    private static String[] parseEnPage(String html) {
        Matcher m = MAIL_LINK.matcher(html);
        if (!m.find()) {
            throw new RuntimeException("anonbox: mailbox link not found");
        }
        String inbox = m.group(1);
        String secret = m.group(2);
        String token = inbox + "/" + secret;

        // 查找邮箱地址
        String addressDisplay = null;
        Matcher ddMatcher = DD_RE.matcher(html);
        while (ddMatcher.find()) {
            String attrs = ddMatcher.group(1);
            String inner = ddMatcher.group(2);
            if (DISPLAY_NONE.matcher(attrs).find()) continue;
            Matcher pm = P_RE.matcher(inner);
            if (!pm.find()) continue;
            String pInner = SPAN_RE.matcher(pm.group(1)).replaceAll("");
            String display = stripTags(pInner);
            if (!display.contains("@")) continue;
            int at = display.lastIndexOf('@');
            String domain = display.substring(at + 1).trim().toLowerCase();
            if ("googlemail.com".equals(domain)) continue;
            String expectedDomain = (inbox + ".anonbox.net").toLowerCase();
            if (!domain.equals(expectedDomain)) continue;
            String local = display.substring(0, at).trim();
            if (local.isEmpty()) continue;
            addressDisplay = display;
            break;
        }
        if (addressDisplay == null) {
            throw new RuntimeException("anonbox: address paragraph not found");
        }

        int at = addressDisplay.indexOf('@');
        String local = addressDisplay.substring(0, at).trim();
        if (local.isEmpty()) {
            throw new RuntimeException("anonbox: empty local part");
        }
        String emailAddr = local + "@" + inbox + ".anonbox.net";

        // 提取过期时间
        Matcher em = EXPIRES_RE.matcher(html);
        String exp = em.find() ? em.group(1).trim() : null;

        return new String[]{emailAddr, token, exp};
    }

    private static Map<String, Object> mboxBlockToRaw(String block, String recipient) {
        String normalized = block.replace("\r\n", "\n");
        String[] lines = normalized.split("\n");
        int i = 0;
        if (lines.length > 0 && lines[0].startsWith("From ")) i = 1;

        // 解析头部
        Map<String, String> headers = new LinkedHashMap<>();
        String curKey = "";
        while (i < lines.length) {
            String line = lines[i];
            if (line.isEmpty()) { i++; break; }
            if ((line.charAt(0) == ' ' || line.charAt(0) == '\t') && !curKey.isEmpty()) {
                headers.put(curKey, headers.get(curKey) + " " + line.trim());
            } else {
                int idx = line.indexOf(':');
                if (idx > 0) {
                    curKey = line.substring(0, idx).trim().toLowerCase();
                    headers.put(curKey, line.substring(idx + 1).trim());
                }
            }
            i++;
        }

        // 解析正文
        StringBuilder bodyBuilder = new StringBuilder();
        while (i < lines.length) {
            if (bodyBuilder.length() > 0) bodyBuilder.append("\n");
            bodyBuilder.append(lines[i]);
            i++;
        }
        String body = bodyBuilder.toString();

        String ct = headers.getOrDefault("content-type", "").toLowerCase();
        String text = "";
        String htmlBody = "";

        if (ct.contains("multipart/")) {
            Matcher bm = Pattern.compile("(?i)boundary=\"?([^\"\\s;]+)\"?").matcher(ct);
            if (bm.find()) {
                String boundary = bm.group(1);
                String[] parts = body.split("--" + Pattern.quote(boundary));
                for (String part : parts) {
                    part = part.trim();
                    if (part.isEmpty() || part.equals("--")) continue;
                    int sep = part.indexOf("\n\n");
                    if (sep < 0) continue;
                    String ph = part.substring(0, sep);
                    String pb = part.substring(sep + 2);
                    String pct = "";
                    Matcher pctM = Pattern.compile("(?i)^content-type:\\s*([^\\s;]+)", Pattern.MULTILINE).matcher(ph);
                    if (pctM.find()) pct = pctM.group(1).toLowerCase();
                    if ("text/plain".equals(pct)) text = pb.trim();
                    else if ("text/html".equals(pct)) htmlBody = pb.trim();
                }
            }
        }
        if (text.isEmpty() && htmlBody.isEmpty()) {
            if (ct.contains("text/html")) htmlBody = body.trim();
            else text = body.trim();
        }

        String dateStr = headers.getOrDefault("date", "");
        String msgId = headers.getOrDefault("message-id", simpleHash(block.substring(0, Math.min(512, block.length()))));

        Map<String, Object> raw = new LinkedHashMap<>();
        raw.put("id", msgId);
        raw.put("from", headers.getOrDefault("from", ""));
        raw.put("to", headers.getOrDefault("to", recipient));
        raw.put("subject", headers.getOrDefault("subject", ""));
        raw.put("text", text);
        raw.put("html", htmlBody);
        raw.put("date", dateStr);
        return raw;
    }

    private static String simpleHash(String s) {
        long h = 0;
        for (int i = 0; i < s.length(); i++) {
            h = (h * 31 + s.charAt(i)) & 0xFFFFFFFFL;
        }
        if (h == 0) return "0";
        String digits = "0123456789abcdefghijklmnopqrstuvwxyz";
        StringBuilder out = new StringBuilder();
        while (h > 0) {
            out.append(digits.charAt((int) (h % 36)));
            h /= 36;
        }
        return out.reverse().toString();
    }
}

File: sdk/kotlin/src/main/kotlin/com/xxxxteam/tempmail/providers/Anonbox.kt
package com.xxxxteam.tempmail.providers

import com.xxxxteam.tempmail.Email
import com.xxxxteam.tempmail.EmailInfo
import com.xxxxteam.tempmail.Normalize
import com.xxxxteam.tempmail.Provider

/**
 * Anonbox 渠道（anonbox.net，CCC）。
 *
 * GET /en/ 解析 HTML 获取邮箱地址与路径令牌 "{inbox}/{secret}"，
 * GET /{inbox}/{secret}/ 返回 mbox 纯文本，逐块解析。
 */
object Anonbox : Provider {

    private const val PAGE_URL = "https://anonbox.net/en/"
    private const val BASE = "https://anonbox.net"

    private val MAIL_LINK = Regex(
        "<a href=\"https://anonbox\\.net/([a-z0-9-]+)/([A-Za-z0-9._~-]+)\">https://anonbox\\.net/[^\"]+</a>",
        RegexOption.IGNORE_CASE)
    private val DD_RE = Regex("<dd([^>]*)>([\\s\\S]*?)</dd>", setOf(RegexOption.IGNORE_CASE, RegexOption.DOT_MATCHES_ALL))
    private val DISPLAY_NONE = Regex("display\\s*:\\s*none", RegexOption.IGNORE_CASE)
    private val P_RE = Regex("<p>([\\s\\S]*?)</p>", setOf(RegexOption.IGNORE_CASE, RegexOption.DOT_MATCHES_ALL))
    private val SPAN_RE = Regex("<span\\b[^>]*>[\\s\\S]*?</span>", setOf(RegexOption.IGNORE_CASE, RegexOption.DOT_MATCHES_ALL))
    private val TAG_RE = Regex("<[^>]+>")
    private val EXPIRES_RE = Regex(
        "Your mail address is valid until:</dt>\\s*<dd><p>([^<]+)</p>",
        setOf(RegexOption.IGNORE_CASE, RegexOption.DOT_MATCHES_ALL))
    private val BLOCK_SPLIT = Regex("(?m)(?=^From )")
    private val BOUNDARY_RE = Regex("boundary=\"?([^\"\\s;]+)\"?", RegexOption.IGNORE_CASE)
    private val PART_CT_RE = Regex("^content-type:\\s*([^\\s;]+)", setOf(RegexOption.IGNORE_CASE, RegexOption.MULTILINE))

    private val HTML_HEADERS = mapOf(
        "User-Agent" to "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 " +
            "(KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36",
        "Accept" to "text/html,application/xhtml+xml",
    )
    private val MBOX_HEADERS = mapOf(
        "User-Agent" to "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 " +
            "(KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36",
        "Accept" to "text/plain,*/*",
    )

    /** 创建临时邮箱（解析 /en/ 页面）。 */
    override suspend fun generate(): EmailInfo {
        val r = ProviderUtil.httpGet(PAGE_URL, HTML_HEADERS)
        r.ensureSuccess()
        val (email, token, exp) = parseEnPage(r.body)
        return EmailInfo(email = email, channel = "anonbox", token = token, expiresAt = exp)
    }

    /** 获取邮件列表（mbox 格式解析）。 */
    override suspend fun getEmails(info: EmailInfo): List<Email> {
        val token = info.token.trim()
        if (token.isEmpty()) throw RuntimeException("anonbox: token 不能为空")
        val path = if (token.endsWith("/")) token else "$token/"
        val r = ProviderUtil.httpGet("$BASE/$path", MBOX_HEADERS)
        if (r.statusCode == 404) return emptyList()
        r.ensureSuccess()

        val raw = r.body.trim()
        if (raw.isEmpty()) return emptyList()
        return BLOCK_SPLIT.split(raw)
            .map { it.trim() }
            .filter { it.isNotEmpty() }
            .map { Normalize.fromMap(mboxBlockToRaw(it, info.email), info.email) }
    }

    /** 解析 /en/ 页面，返回 (邮箱, 路径令牌, 过期时间)。 */
    private fun parseEnPage(html: String): Triple<String, String, String> {
        val link = MAIL_LINK.find(html) ?: throw RuntimeException("anonbox: 未找到邮箱链接")
        val inbox = link.groupValues[1]
        val secret = link.groupValues[2]
        val token = "$inbox/$secret"

        var addressDisplay: String? = null
        for (dd in DD_RE.findAll(html)) {
            val attrs = dd.groupValues[1]
            val inner = dd.groupValues[2]
            if (DISPLAY_NONE.containsMatchIn(attrs)) continue
            val pm = P_RE.find(inner) ?: continue
            val pInner = SPAN_RE.replace(pm.groupValues[1], "")
            val display = stripTags(pInner)
            if (!display.contains("@")) continue
            val at = display.lastIndexOf('@')
            val domain = display.substring(at + 1).trim().lowercase()
            if (domain == "googlemail.com") continue
            if (domain != "$inbox.anonbox.net".lowercase()) continue
            if (display.substring(0, at).trim().isEmpty()) continue
            addressDisplay = display
            break
        }
        if (addressDisplay == null) throw RuntimeException("anonbox: 未找到邮箱地址段落")

        val at = addressDisplay.indexOf('@')
        val local = addressDisplay.substring(0, at).trim()
        if (local.isEmpty()) throw RuntimeException("anonbox: 邮箱本地部分为空")
        val emailAddr = "$local@$inbox.anonbox.net"

        val exp = EXPIRES_RE.find(html)?.groupValues?.get(1)?.trim() ?: ""
        return Triple(emailAddr, token, exp)
    }

    /** 将单个 mbox 块解析为标准字段字典。 */
    private fun mboxBlockToRaw(block: String, recipient: String): Map<String, Any?> {
        val normalized = block.replace("\r\n", "\n")
        val lines = normalized.split("\n")
        var i = 0
        if (lines.isNotEmpty() && lines[0].startsWith("From ")) i = 1

        val headers = LinkedHashMap<String, String>()
        var curKey = ""
        while (i < lines.size) {
            val line = lines[i]
            if (line.isEmpty()) { i++; break }
            if ((line[0] == ' ' || line[0] == '\t') && curKey.isNotEmpty()) {
                headers[curKey] = headers[curKey] + " " + line.trim()
            } else {
                val idx = line.indexOf(':')
                if (idx > 0) {
                    curKey = line.substring(0, idx).trim().lowercase()
                    headers[curKey] = line.substring(idx + 1).trim()
                }
            }
            i++
        }

        val bodyBuilder = StringBuilder()
        while (i < lines.size) {
            if (bodyBuilder.isNotEmpty()) bodyBuilder.append("\n")
            bodyBuilder.append(lines[i])
            i++
        }
        val body = bodyBuilder.toString()

        val ct = (headers["content-type"] ?: "").lowercase()
        var text = ""
        var htmlBody = ""
        if (ct.contains("multipart/")) {
            val boundary = BOUNDARY_RE.find(ct)?.groupValues?.get(1)
            if (boundary != null) {
                val parts = body.split("--" + Regex.escape(boundary).let { Regex(it) })
                for (rawPart in parts) {
                    val part = rawPart.trim()
                    if (part.isEmpty() || part == "--") continue
                    val sep = part.indexOf("\n\n")
                    if (sep < 0) continue
                    val ph = part.substring(0, sep)
                    val pb = part.substring(sep + 2)
                    val pct = PART_CT_RE.find(ph)?.groupValues?.get(1)?.lowercase() ?: ""
                    when (pct) {
                        "text/plain" -> text = pb.trim()
                        "text/html" -> htmlBody = pb.trim()
                    }
                }
            }
        }
        if (text.isEmpty() && htmlBody.isEmpty()) {
            if (ct.contains("text/html")) htmlBody = body.trim() else text = body.trim()
        }

        val msgId = headers["message-id"]
            ?: simpleHash(block.substring(0, minOf(512, block.length)))
        return mapOf(
            "id" to msgId,
            "from" to (headers["from"] ?: ""),
            "to" to (headers["to"] ?: recipient),
            "subject" to (headers["subject"] ?: ""),
            "text" to text,
            "html" to htmlBody,
            "date" to (headers["date"] ?: ""),
        )
    }

    /** 去除 HTML 标签并解码常见实体。 */
    private fun stripTags(html: String): String =
        TAG_RE.replace(html, "")
            .replace("&nbsp;", " ").replace("&amp;", "&")
            .replace("&lt;", "<").replace("&gt;", ">").trim()

    /** 简单 36 进制哈希，用于缺失 Message-Id 时生成稳定 id。 */
    private fun simpleHash(s: String): String {
        var h = 0L
        for (c in s) h = (h * 31 + c.code) and 0xFFFFFFFFL
        if (h == 0L) return "0"
        val digits = "0123456789abcdefghijklmnopqrstuvwxyz"
        val out = StringBuilder()
        while (h > 0) {
            out.append(digits[(h % 36).toInt()])
            h /= 36
        }
        return out.reverse().toString()
    }
}
